Audio subject

hughdickinson opened the pull request:
Fixes #3809 .

Implemented audio subjects for users with Admin access using a similar approach to that used for video subjects.

Replicated the tests that are allocated to video subjects.

Example project staged [here](https://audio-subject.pfe-preview.zooniverse.org/projects/hughdickinson/testing-audio-subjects).

# Review Checklist

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [ ] Did you deploy a staging branch?
- [ ] Did you get any type checking errors from [Babel Typecheck](https://github.com/codemix/babel-plugin-typecheck)


## Optional

- [ ] If it's a new component, is it in ES6? Is it clear of warnings from ESLint?
- [ ] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [ ] If changes are made to the classifier, does the dev classifier still work?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?

camallen commented on app/pages/lab/subject-set.cjsx:
```diff
@@ -13,9 +13,9 @@ isAdmin = require '../../lib/is-admin'
+VALID_SUBJECT_EXTENSIONS = ['.jpg', '.jpeg', '.png', '.gif', '.svg', '.mp3']
 INVALID_FILENAME_CHARS = ['/', '\\', ':', ',']
-MAX_FILE_SIZE = 600 * 1024
+MAX_FILE_SIZE = 4000 * 1024
```
Do we really want to bump the file size this much? 4MB seems too big for downloading with preload to auto, especially on mobile devices.

eatyourgreens replied:
This does seem excessive since it would correspond to a few minutes of audio playback, I think.

camallen commented on app/components/file-viewer/audio-player.jsx:
```diff
@@ -0,0 +1,80 @@
+          ref={(element) => { this.player = element; }}
+          src={this.props.src}
+          type={`${this.props.type}/${this.props.format}`}
+          preload="auto"
```
This is only a browser hint but should this be auto? Considering asset sizes it may be best to avoid automatic download of the media asset unless the user presses play. 

'metadata' might be a better fit, https://developer.mozilla.org/en/docs/Web/HTML/Element/audio

eatyourgreens replied:
From memory, `autoplay="metadata"` never triggers the `canplay` event in some browsers, so I'd be cautious about changing this. See #2162

camallen replied:
Be nice to test this again @hughdickinson - optimal solution is to not download automatically (avoid bandwidth costs on client / server sides) but have the media resource playable ;)

eatyourgreens replied:
Modern browsers are pretty good about conserving bandwidth, particularly on mobile. `preload="auto"` shouldn't download the whole file, just enough to start playing when the play button is pressed, unless `autoplay` is set.

camallen replied:
i couldn't see this documented anywhere when reading this morning, happy to defer to your browser knowledge. 

We should be aware and thinking about this issue as I know i'd be annoyed if my quota was used up downloading large media resources.

eatyourgreens replied:
I believe each browser has its own idea of how to implement the Media API. Fun times. ☹️ 

In Chrome, at least, you can check the GET request for the Arizona Batwatch mp4 files in the network tab and, I think, only the first portion of the file is downloaded.

camallen replied:
in my chrome & firefox i get 3 mp4 media requests, the first two small portions the third retrieves the remaining content all just on page load.

eatyourgreens commented on app/components/file-viewer/audio-player.jsx:
```diff
@@ -0,0 +1,80 @@
+          Your browser does not support the audio format. Please upgrade your browser.
+        </audio>
+
+        {this.props.showControls}
```
`showControls` is a boolean. Why is it being rendered here?

eatyourgreens commented on app/components/file-viewer/audio-player.jsx:
```diff
@@ -0,0 +1,80 @@
+};
+
+AudioPlayer.defaultProps = {
+  showControls: true
```
I think you can get rid of this. `showControls` is used to render custom controls in the video player, but there are no custom controls in this component.

eatyourgreens commented on app/components/file-viewer/audio-player.jsx:
```diff
@@ -0,0 +1,80 @@
+  }
+
+  render() {
+    // const rates = [0.25, 0.5, 1];
```
Delete this line?

eatyourgreens commented on app/lib/get-subject-location.coffee:
```diff
@@ -1,6 +1,7 @@
 READABLE_FORMATS =
   image: ['jpeg', 'png', 'svg+xml', 'gif']
   video: ['mp4']
+  audio: ['mpeg']
```
Probably not important, but this doesn't match the MIME type used for the test subject.

eatyourgreens commented:
I'm getting an error for the staging project.
`Please try again. Something went wrong: Error: no subject set is associated with this workflow`

hughdickinson commented:
I've implemented the suggested changes and fixed the subject set on the demo project. Sorry about the latter, I was trying to use the same project to test an audio+image branch I'm working on!

eatyourgreens commented:
👍 
One small thing: If I favourite an audio file, there's no preview defined for collections, but I don't think that's a blocking issue.

eatyourgreens commented:
We should probably also think about how audio files are displayed in Talk. https://audio-subject.pfe-preview.zooniverse.org/projects/hughdickinson/testing-audio-subjects/talk/subjects/50246

srallen commented:
Videos don't have previews either. As we expand on non-image subject types, we'll need to re-think the subject viewer previews in talk and in collections.

eatyourgreens commented:
Videos render a `video` tag for [Talk previews](https://www.zooniverse.org/projects/zooniverse/arizona-batwatch/talk/230). I think it's defined in the `Thumbnail` component somewhere.

There's also a [larger view for hashtag lists](https://www.zooniverse.org/projects/zooniverse/arizona-batwatch/talk/tags/insect), but I think that uses the standard subject viewer.

srallen commented:
Ah right, then it's just a problem for collections: https://www.zooniverse.org/projects/zooniverse/arizona-batwatch/collections

eatyourgreens commented:
Yeah, they might be using CSS background images.

hughdickinson commented:
Can I just check whether I need to make additional changes. If not will somebody else take care of actually doing the merge?
